crates/bsp_types/src/extensions/rust/rust_workspace.rs

```diff
@@ -16,141 +16,236 @@ impl Request for RustWorkspace {
-    pub all_targets: Vec<RustTarget>,
+    /** Correspond to source files which can be compiled into a crate from this package.
+    Contains only resolved targets without conflicts. */
+    pub targets: Vec<RustBuildTarget>,
+    /** Same as `targets`, but contains all targets from this package.
+    `targets` should be the subset of `allTargets`. */
+    pub all_targets: Vec<RustBuildTarget>,
+    /** Set of features defined for the package (including optional dependencies).
+    Each feature maps to an array of features or dependencies it enables.
+    The entry named "default" defines which features are enabled by default. */
     pub features: Vec<RustFeature>,
-    pub enabled_features: Vec<String>, //?
+    /** Array of features enabled on this package. */
+    pub enabled_features: Vec<String>,
+    /** Conditional compilation flags that can be set based on certain conditions.
+    They can be used to enable or disable certain sections of code during the build process. */
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub cfg_options: Option<RustCfgOptions>, //Null or check where it comes from in current plugin implementaion
-    pub env: HashMap<String, String>, //? to co ma plugin: https://github.com/intellij-rust/intellij-rust/blob/d99a5fcd5de6dd4bd81d18d67e0c6718e7612127/src/main/kotlin/org/rust/cargo/toolchain/impl/CargoMetadata.kt#L438 to co wysyła ZPP: https://github.com/ZPP-This-is-fine/bazel-bsp/blob/712e005abcd9d3f0a02a2d2001d486f2c728559e/server/src/main/java/org/jetbrains/bsp/bazel/server/sync/languages/rust/RustWorkspaceResolver.kt#L155
+    pub cfg_options: Option<RustCfgOptions>,
+    /** Environment variables for the package. */
+    pub env: HashMap<String, String>,
+    /** An absolute path which is used as a value of `OUT_DIR` environmental
+    variable when compiling current package. */
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub out_dir_url: Option<String>, // tutaj Null, bo nie mamy pojęcia co to
+    pub out_dir_url: Option<Uri>,
+    /** File path to compiled output of a procedural macro crate.
+    Procedural macros are macros that generate code at compile time.
+    Contains files with file extensions: `.dll`, `.so` or `.dylib`. */
     #[serde(skip_serializing_if = "Option::is_none")]
-    pub proc_macro_artifact: Option<RustProcMacroArtifact>, //?
+    pub proc_macro_artifact: Option<String>,
 }
 
 #[derive(Serialize, Deserialize, Default)]
 #[serde(rename_all = "camelCase")]
-pub struct DepKind {
-    pub kind: DepKindEnum,
+pub struct RustDepKindInfo {
+    /** The dependency kind. */
+    pub kind: RustDepKind,
+    /** The target platform for the dependency. */
     #[serde(skip_serializing_if = "Option::is_none")]
     pub target: Option<String>,
 }
 
 #[derive(Serialize_repr, Deserialize_repr, Default)]
 #[repr(u8)]
-pub enum DepKindEnum {
+pub enum RustDepKind {
+    /** For old Cargo versions prior to `1.41.0`. */
     Unclassified = 1,
-    Stdlib = 2,
```

Comment:
czy to się zmieniło w cargo metadata? bo jak to się może tak zmieniać to nie powinniśmy robić closed enuma...

Reply:
W cargo się nic nie zmieniła. Ta zmiana wynikała z tego, że poprzedni team przepisał te kind na podstawie tego jak są one określone w intellij-rust, jednak on dodaje sobie kind stdlib albo do testów albo żeby oznaczać dociągane stdlib (które nawet przy korzystaniu z bsp jest brane bezpośrednio w pluginie intellij-rust). W naszym przypadku natomiast nie powinno być takiego kind, nie wspomina też o nim w ogóle dokumentacja `cargo metadata` od której je dostajemy. Natomiast jeżeli uważasz że bezpieczniej byłoby to zamienić na open string enum to mogę tak zrobić

Reply:
Zmienione